From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Tariq Bashir <120014322+t-bashir-bs@users.noreply.github.com>
Date: Tue, 24 Sep 2024 13:44:25 +0100
Subject: OS-17652: Add support for window rotation

This patch introduces a new WebPreference and browser window API method
setWindowTransform to control window rotation
(90, 180, or 270 degrees) in BrightSign. The preference is set in
WebContentsImpl and passed to RenderWidgetHostViewAura, which then
applies the display transform hint on WindowTreeHost if rotation is
requested. Method setWindowTransform also applies the display transform
hint on WindowTreeHost if rotation is requested.

Method WindowTreeHost::GetTransformedWindowBounds was added to compute the
transformed bounds when DesktopNativeWidgetAura::SetBounds is invoked.

Method WindowTreeHost::GetTransformedCompositorBounds was added to compute
the transformed bounds when Window::SetBoundsInternal and
WindowTreeHost::GetTransformedCompositorBounds are invoked. This method
handles the transformed bounds for both popups and windows. Since window
transforms have the same functionality for both compositor and window bounds,
the method uses GetTransformedWindowBounds to compute the transformed bounds
for windows.

For windows these transform methods ensure bounds are transposed only once
by checking the aspect ratio of the current and new bounds, avoiding
multiple transpositions. This aspect ratio comparison approach is more robust
than relying on the original bounds, as Window::SetBoundsInternal may receive
bounds adjusted for padding or insets.
For popups we get the initial physical bounds during the initialisation of
the popup and so use this directly to calculate the transformed bounds, rather
than comparing aspect ratios.

Method GetTransformedPopupBounds is used to compute the transformed bounds
of the physical popup window. This method is called in
RenderWidgetHostViewAura::InitAsPopup and RenderWidgetHostViewAura::SetBounds.
The transform used by this method is fetched from the parent RenderWidgetHostView
as this transform is based on the parent screen dimensions.

For input event transforms:
When GetTransformedWindowBounds is invoked, it also computes an input
transform based on the display transform hint, which is subsequently
applied to mouse and touch events in WindowEventDispatcher::PreDispatchEvent.
The non_transformed_bounds method, added to Window, stores the original bounds,
enabling WindowTargeter to align hit-test rectangles with screen resolution,
thus ensuring consistent mouse and touch input alignment.

diff --git a/components/viz/service/display/output_surface.h b/components/viz/service/display/output_surface.h
index 56197a333ce17f0bdcbf766e3e3e8b1bf459a479..edb531fff186c6a48ba13355c11c050796de7964 100644
--- a/components/viz/service/display/output_surface.h
+++ b/components/viz/service/display/output_surface.h
@@ -97,7 +97,7 @@ class VIZ_SERVICE_EXPORT OutputSurface {
     // to apply resize optimization by padding to its width/height.
     bool supports_viewporter = false;
     // OutputSurface's orientation mode.
-    OrientationMode orientation_mode = OrientationMode::kLogic;
+    OrientationMode orientation_mode = OrientationMode::kHardware;
     // Whether this OutputSurface supports direct composition layers.
     DCSupportLevel dc_support_level = DCSupportLevel::kNone;
     // Whether this OutputSurface should skip DrawAndSwap(). This is true for
diff --git a/components/viz/service/display_embedder/software_output_surface.cc b/components/viz/service/display_embedder/software_output_surface.cc
index 18a7adc8925ff7bf90c4143c564458e7ea9f9008..b5af2405a4cff3865dd386d09c458bb74d1dae6d 100644
--- a/components/viz/service/display_embedder/software_output_surface.cc
+++ b/components/viz/service/display_embedder/software_output_surface.cc
@@ -138,8 +138,13 @@ void SoftwareOutputSurface::SetUpdateVSyncParametersCallback(
   update_vsync_parameters_callback_ = std::move(callback);
 }
 
+void SoftwareOutputSurface::SetDisplayTransformHint(
+    gfx::OverlayTransform transform) {
+  display_transform_ = transform;
+}
+
 gfx::OverlayTransform SoftwareOutputSurface::GetDisplayTransform() {
-  return gfx::OVERLAY_TRANSFORM_NONE;
+  return display_transform_;
 }
 
 #if BUILDFLAG(IS_LINUX)
diff --git a/components/viz/service/display_embedder/software_output_surface.h b/components/viz/service/display_embedder/software_output_surface.h
index 014a17a1d2326a6d7062233d06edf1d75962f76d..6eabf0892695c10726437b4bd808a38c4b0a03dd 100644
--- a/components/viz/service/display_embedder/software_output_surface.h
+++ b/components/viz/service/display_embedder/software_output_surface.h
@@ -39,7 +39,7 @@ class VIZ_SERVICE_EXPORT SoftwareOutputSurface : public OutputSurface {
   void SwapBuffers(OutputSurfaceFrame frame) override;
   void SetUpdateVSyncParametersCallback(
       UpdateVSyncParametersCallback callback) override;
-  void SetDisplayTransformHint(gfx::OverlayTransform transform) override {}
+  void SetDisplayTransformHint(gfx::OverlayTransform transform) override;
   gfx::OverlayTransform GetDisplayTransform() override;
 #if BUILDFLAG(IS_LINUX)
   void SetNeedsSwapSizeNotifications(
@@ -63,6 +63,8 @@ class VIZ_SERVICE_EXPORT SoftwareOutputSurface : public OutputSurface {
   bool needs_swap_size_notifications_ = false;
 #endif
 
+  gfx::OverlayTransform display_transform_ = gfx::OVERLAY_TRANSFORM_NONE;
+
   base::WeakPtrFactory<SoftwareOutputSurface> weak_factory_{this};
 };
 
diff --git a/content/browser/renderer_host/render_widget_host_view_aura.cc b/content/browser/renderer_host/render_widget_host_view_aura.cc
index d0e94537ef0e03cb248d6c35b6dde851c52cd6e4..26f22c824d3f7243e6d13839ccb6df7d58772540 100644
--- a/content/browser/renderer_host/render_widget_host_view_aura.cc
+++ b/content/browser/renderer_host/render_widget_host_view_aura.cc
@@ -126,6 +126,7 @@
 #include "ui/base/ime/linux/text_edit_command_auralinux.h"
 #include "ui/base/ime/text_input_flags.h"
 #include "ui/linux/linux_ui.h"
+#include "ui/ozone/platform_selection.h"
 #endif
 
 #if BUILDFLAG(IS_CHROMEOS)
@@ -308,8 +309,23 @@ RenderWidgetHostViewAura::RenderWidgetHostViewAura(
             .double_tap_to_zoom_enabled;
 
     event_handler_->SetPinchZoomEnabled(owner_delegate->GetWebkitPreferencesForWidget().enable_pinch_zoom);
-  }
 
+    switch (owner_delegate->GetWebkitPreferencesForWidget().window_transform_type)
+    {
+      case blink::mojom::WindowTransformType::kWindowTransformTypeRotate90:
+        window_transform_ = gfx::OVERLAY_TRANSFORM_ROTATE_CLOCKWISE_90;
+        break;
+      case blink::mojom::WindowTransformType::kWindowTransformTypeRotate180:
+        window_transform_ = gfx::OVERLAY_TRANSFORM_ROTATE_CLOCKWISE_180;
+        break;
+      case blink::mojom::WindowTransformType::kWindowTransformTypeRotate270:
+        window_transform_ = gfx::OVERLAY_TRANSFORM_ROTATE_CLOCKWISE_270;
+        break;
+      default:
+        window_transform_ = gfx::OVERLAY_TRANSFORM_NONE;
+        break;
+    }
+  }
   host()->render_frame_metadata_provider()->AddObserver(this);
 }
 
@@ -332,6 +348,11 @@ void RenderWidgetHostViewAura::InitAsChild(gfx::NativeView parent_view) {
       UpdateSystemCursorSize(cursor_client->GetSystemCursorSize());
   }
 
+  aura::WindowTreeHost* host = window_->GetHost();
+  if (host && window_transform_ != gfx::OVERLAY_TRANSFORM_NONE) {
+    host->SetDisplayTransformHint(window_transform_);
+  }
+
 #if BUILDFLAG(IS_WIN)
   // This will fetch and set the display features.
   ObserveDevicePosturePlatformProvider();
@@ -389,14 +410,31 @@ void RenderWidgetHostViewAura::InitAsPopup(
         popup_parent_host_view_->window_, window_);
   }
 
+  window_transform_ = popup_parent_host_view_->GetDisplayTransform();
+  gfx::Rect final_bounds_in_screen(bounds_in_screen);
+  aura::WindowTreeHost* parent_host = popup_parent_host_view_->window_->GetHost();
+  if (parent_host) {
+    original_popup_bounds_ = bounds_in_screen;
+    final_bounds_in_screen = GetTransformedPopupBounds(
+         bounds_in_screen, parent_host->GetBoundsInPixels(), parent_host->GetPopupTransform());
+  }
+
+  // Disable using anchor rect for all platforms as when using it with Wayland it is
+  // not positioning pop ups in the right place when they are near to edges of the
+  // screen (they will get clipped). Also, Nexus doesn't use it.
+  // This will result in us having the same behavior as when using Nexus and also
+  // mean we don't have to add additional functionality to transform the anchor rect
+  // when the html window is rotated.
+#if 0
   ui::OwnedWindowAnchor owned_window_anchor = {
       anchor_rect, ui::OwnedWindowAnchorPosition::kBottomLeft,
       ui::OwnedWindowAnchorGravity::kBottomRight,
       ui::OwnedWindowConstraintAdjustment::kAdjustmentFlipY};
   window_->SetProperty(aura::client::kOwnedWindowAnchor, owned_window_anchor);
+#endif
 
   aura::Window* root = popup_parent_host_view_->window_->GetRootWindow();
-  aura::client::ParentWindowWithContext(window_, root, bounds_in_screen,
+  aura::client::ParentWindowWithContext(window_, root, final_bounds_in_screen,
                                         display::kInvalidDisplayId);
 
   SetBounds(bounds_in_screen);
@@ -420,6 +458,11 @@ void RenderWidgetHostViewAura::InitAsPopup(
   if (cursor_client)
     UpdateSystemCursorSize(cursor_client->GetSystemCursorSize());
 
+  aura::WindowTreeHost* host = window_->GetHost();
+  if (host) {
+    host->SetPopupBounds(bounds_in_screen);
+    host->SetDisplayTransformHint(window_transform_);
+  }
 #if BUILDFLAG(IS_WIN)
   // This will fetch and set the display features.
   ObserveDevicePosturePlatformProvider();
@@ -440,7 +483,14 @@ void RenderWidgetHostViewAura::SetSize(const gfx::Size& size) {
 }
 
 void RenderWidgetHostViewAura::SetBounds(const gfx::Rect& rect) {
-  gfx::Point relative_origin(rect.origin());
+  gfx::Rect transformed_rect(rect);
+  aura::WindowTreeHost* parent_host = popup_parent_host_view_->window_->GetHost();;
+  if (parent_host) {
+    transformed_rect = GetTransformedPopupBounds(
+        rect, parent_host->GetBoundsInPixels(), parent_host->GetPopupTransform());
+  }
+
+  gfx::Point relative_origin(transformed_rect.origin());
 
   // RenderWidgetHostViewAura::SetBounds() takes screen coordinates, but
   // Window::SetBounds() takes parent coordinates, so do the conversion here.
@@ -454,7 +504,7 @@ void RenderWidgetHostViewAura::SetBounds(const gfx::Rect& rect) {
     }
   }
 
-  InternalSetBounds(gfx::Rect(relative_origin, rect.size()));
+  InternalSetBounds(gfx::Rect(relative_origin, transformed_rect.size()));
 }
 
 gfx::NativeView RenderWidgetHostViewAura::GetNativeView() {
@@ -3372,4 +3422,49 @@ ui::Compositor* RenderWidgetHostViewAura::GetCompositor() {
   return window_->GetHost()->compositor();
 }
 
+gfx::Rect RenderWidgetHostViewAura::GetTransformedPopupBounds(const gfx::Rect& new_bounds, const gfx::Rect& parent_bounds, const gfx::Transform& parent_popup_transform) {
+  gfx::Rect transformed_bounds(new_bounds);
+
+  // Wayland draws the popup window in the parent window's coordinate space. However, X11 (for CI)
+  // and Nexus draw the popup window in the root window's coordinate space.
+  // This means we need to remove the parent window offset, then apply the parent_popup_transform,
+  // and then add the parent window offset back for platforms other than wayland.
+  transformed_bounds.Offset(-parent_bounds.OffsetFromOrigin());
+  transformed_bounds.set_origin(parent_popup_transform.MapPoint(transformed_bounds.origin()));
+#if BUILDFLAG(IS_LINUX)
+  if (std::string("wayland") != ui::GetOzonePlatformName())
+#endif
+    transformed_bounds.Offset(parent_bounds.OffsetFromOrigin());
+
+  switch (window_transform_) {
+    case gfx::OVERLAY_TRANSFORM_ROTATE_CLOCKWISE_90:
+      transformed_bounds.set_size(gfx::Size(original_popup_bounds_.height(), original_popup_bounds_.width()));
+      transformed_bounds.Offset(-original_popup_bounds_.height(), 0);
+      break;
+
+    case gfx::OVERLAY_TRANSFORM_ROTATE_CLOCKWISE_180:
+      transformed_bounds.Offset(-original_popup_bounds_.width(), -original_popup_bounds_.height());
+      break;
+
+    case gfx::OVERLAY_TRANSFORM_ROTATE_CLOCKWISE_270:
+      transformed_bounds.set_size(gfx::Size(original_popup_bounds_.height(), original_popup_bounds_.width()));
+      transformed_bounds.Offset(0, -original_popup_bounds_.width());
+      break;
+
+    default:
+      break;
+  }
+  return transformed_bounds;
+}
+
+gfx::OverlayTransform RenderWidgetHostViewAura::GetDisplayTransform() {
+  gfx::OverlayTransform transform = gfx::OVERLAY_TRANSFORM_NONE;
+
+  aura::WindowTreeHost* host = window_->GetHost();
+  if (host) {
+    transform = host->GetDisplayTransformHint();
+  }
+  return transform;
+}
+
 }  // namespace content
diff --git a/content/browser/renderer_host/render_widget_host_view_aura.h b/content/browser/renderer_host/render_widget_host_view_aura.h
index c54ed7d8f5d4d371626865c6ae63ef2efbef1dba..e3bc171200f73604f505af8e1219532e8f763205 100644
--- a/content/browser/renderer_host/render_widget_host_view_aura.h
+++ b/content/browser/renderer_host/render_widget_host_view_aura.h
@@ -47,6 +47,7 @@
 #include "ui/display/display_observer.h"
 #include "ui/gfx/geometry/insets.h"
 #include "ui/gfx/geometry/rect.h"
+#include "ui/gfx/overlay_transform.h"
 #include "ui/gfx/selection_bound.h"
 #include "ui/wm/public/activation_delegate.h"
 
@@ -440,6 +441,10 @@ class CONTENT_EXPORT RenderWidgetHostViewAura
 
   ui::Compositor* GetCompositor() override;
 
+  gfx::Rect GetTransformedPopupBounds(
+      const gfx::Rect& new_bounds, const gfx::Rect& parent_bounds, const gfx::Transform& parent_popup_transform);
+  gfx::OverlayTransform GetDisplayTransform();
+
   DelegatedFrameHost* GetDelegatedFrameHostForTesting() const {
     return delegated_frame_host_.get();
   }
@@ -876,6 +881,9 @@ class CONTENT_EXPORT RenderWidgetHostViewAura
 
   std::optional<display::ScopedDisplayObserver> display_observer_;
 
+  gfx::OverlayTransform window_transform_ = gfx::OVERLAY_TRANSFORM_NONE;
+  gfx::Rect original_popup_bounds_;
+
   base::WeakPtrFactory<RenderWidgetHostViewAura> weak_ptr_factory_{this};
 };
 
diff --git a/third_party/blink/public/common/web_preferences/web_preferences.h b/third_party/blink/public/common/web_preferences/web_preferences.h
index 90ea9a8fa9ee465ad200dc355ada4474e90c245b..eb9040dfd0814ec3eef44b8fd99c3ec5a927011b 100644
--- a/third_party/blink/public/common/web_preferences/web_preferences.h
+++ b/third_party/blink/public/common/web_preferences/web_preferences.h
@@ -444,6 +444,8 @@ struct BLINK_COMMON_EXPORT WebPreferences {
   // blocking user's access to the background web content.
   bool modal_context_menu = true;
 
+  blink::mojom::WindowTransformType window_transform_type = blink::mojom::WindowTransformType::kWindowTransformTypeNone;
+
   // Whether the safe-area-insets should be changed dynamically based on
   // browser controls shown ratio. This value is used in web settings only
   // when feature DynamicSafeAreaInsets is enabled.
diff --git a/third_party/blink/public/common/web_preferences/web_preferences_mojom_traits.h b/third_party/blink/public/common/web_preferences/web_preferences_mojom_traits.h
index 63b035046ac3d565e5ee1f3a3c559da88d81f059..9f1276bd0eb86f55e5f0a0a9a6bad67029fc411d 100644
--- a/third_party/blink/public/common/web_preferences/web_preferences_mojom_traits.h
+++ b/third_party/blink/public/common/web_preferences/web_preferences_mojom_traits.h
@@ -858,6 +858,11 @@ struct BLINK_COMMON_EXPORT StructTraits<blink::mojom::WebPreferencesDataView,
     return r.payment_request_enabled;
   }
 
+  static blink::mojom::WindowTransformType window_transform_type(
+      const blink::web_pref::WebPreferences& r) {
+    return r.window_transform_type;
+  }
+
   static bool Read(blink::mojom::WebPreferencesDataView r,
                    blink::web_pref::WebPreferences* out);
 };
diff --git a/third_party/blink/public/mojom/webpreferences/web_preferences.mojom b/third_party/blink/public/mojom/webpreferences/web_preferences.mojom
index 142dae3834fe96f120727ef9ffe1250607bcadec..3697c3e4b43b291024785f32dfb2d2ac3482f338 100644
--- a/third_party/blink/public/mojom/webpreferences/web_preferences.mojom
+++ b/third_party/blink/public/mojom/webpreferences/web_preferences.mojom
@@ -109,6 +109,13 @@ enum EffectiveConnectionType {
   kEffectiveConnectionTypeLast,
 };
 
+enum WindowTransformType {
+  kWindowTransformTypeNone,
+  kWindowTransformTypeRotate90,
+  kWindowTransformTypeRotate180,
+  kWindowTransformTypeRotate270,
+};
+
 struct WebPreferences {
   map<string, mojo_base.mojom.String16> standard_font_family_map;
   map<string, mojo_base.mojom.String16> fixed_font_family_map;
@@ -516,4 +523,7 @@ struct WebPreferences {
   // Whether PaymentRequest is enabled. Controlled by WebView settings on
   // WebView and by `kWebPayments` feature flag everywhere.
   bool payment_request_enabled = false;
+
+  // Whether the window should be transformed.
+  WindowTransformType window_transform_type = kWindowTransformTypeNone;
 };
diff --git a/ui/aura/window.cc b/ui/aura/window.cc
index da41875dd57b96745e6721e83a36999bcb2901b3..9a08739a7d9f415a88b90b829c0ace2c2b2117ae 100644
--- a/ui/aura/window.cc
+++ b/ui/aura/window.cc
@@ -1070,10 +1070,16 @@ bool Window::HitTest(const gfx::Point& local_point) {
 
 void Window::SetBoundsInternal(const gfx::Rect& new_bounds) {
   gfx::Rect old_bounds = GetTargetBounds();
+  gfx::Rect final_bounds(new_bounds);
+  WindowTreeHost* host = GetHost();
+  if (host) {
+    // The ui:Layer bounds that we will be setting should match the compositor bounds.
+    final_bounds = host->GetTransformedCompositorBounds(new_bounds);
+  }
 
   // Always need to set the layer's bounds -- even if it is to the same thing.
   // This may cause important side effects such as stopping animation.
-  layer()->SetBounds(new_bounds);
+  layer()->SetBounds(final_bounds);
 
   // If we are currently not the layer's delegate, we will not get bounds
   // changed notification from the layer (this typically happens after animating
@@ -1540,6 +1546,11 @@ void Window::OnLayerBoundsChanged(const gfx::Rect& old_bounds,
   WindowOcclusionTracker::ScopedPause pause_occlusion_tracking;
 
   bounds_ = layer()->bounds();
+  non_transformed_bounds_ = bounds_;
+  WindowTreeHost* host = GetHost();
+  if (host) {
+    non_transformed_bounds_ = host->GetNonTransformedWindowBounds(bounds_);
+  }
 
   if (!IsRootWindow() && old_bounds.size() != bounds_.size() &&
       IsEmbeddingExternalContent()) {
diff --git a/ui/aura/window.h b/ui/aura/window.h
index 96150d989bb99bb151d6a2141d2b04b1549a9dcd..51c04bddf4de4576c6577549a5542c5af72339f9 100644
--- a/ui/aura/window.h
+++ b/ui/aura/window.h
@@ -201,6 +201,7 @@ class AURA_EXPORT Window : public ui::LayerDelegate,
   const WindowDelegate* delegate() const { return delegate_; }
 
   const gfx::Rect& bounds() const { return bounds_; }
+  const gfx::Rect& non_transformed_bounds() const { return non_transformed_bounds_; }
 
   Window* parent() { return parent_; }
   const Window* parent() const { return parent_; }
@@ -742,6 +743,11 @@ class AURA_EXPORT Window : public ui::LayerDelegate,
   // is relative to the parent Window.
   gfx::Rect bounds_;
 
+  // The bounds_ of the layer will be transformed by the target transform. However,
+  // certain operations, such as hit testing, require the bounds_ to be in the
+  // non-transformed state. This is the non-transformed bounds_.
+  gfx::Rect non_transformed_bounds_;
+
   raw_ptr<WindowTreeHost> host_ = nullptr;
 
   client::WindowType type_;
diff --git a/ui/aura/window_event_dispatcher.cc b/ui/aura/window_event_dispatcher.cc
index 67f23e34cc64c6bef6c300b8d031fdf959430c07..98129156d83272321cd0026951e97b1745eb06c9 100644
--- a/ui/aura/window_event_dispatcher.cc
+++ b/ui/aura/window_event_dispatcher.cc
@@ -567,6 +567,14 @@ ui::EventDispatchDetails WindowEventDispatcher::PreDispatchEvent(
     return details;
   }
 
+  // Only convert the location for mouse and touch events as other events, such as gesture, scroll and pinch,
+  // are derived from these and so don't need their locations transformed.
+  if (event->IsMouseEvent() || event->IsTouchEvent()) {
+    ui::LocatedEvent* located_event = event->AsLocatedEvent();
+    located_event->set_location(host_->GetTransformedPoint(located_event->location()));
+    located_event->set_root_location(host_->GetTransformedPoint(located_event->root_location()));
+  }
+
   DispatchDetails details;
   if (event->IsMouseEvent()) {
     details = PreDispatchMouseEvent(target_window, event->AsMouseEvent());
diff --git a/ui/aura/window_targeter.cc b/ui/aura/window_targeter.cc
index 75da6f2dc927d27b0d9944e443c7b636d4c821a6..3442814163ac9b7b908121f4b6e7fa7195bc7060 100644
--- a/ui/aura/window_targeter.cc
+++ b/ui/aura/window_targeter.cc
@@ -39,7 +39,7 @@ bool WindowTargeter::GetHitTestRects(Window* window,
                                      gfx::Rect* hit_test_rect_touch) const {
   DCHECK(hit_test_rect_mouse);
   DCHECK(hit_test_rect_touch);
-  *hit_test_rect_mouse = *hit_test_rect_touch = window->bounds();
+  *hit_test_rect_mouse = *hit_test_rect_touch = window->non_transformed_bounds();
 
   if (ShouldUseExtendedBounds(window)) {
     hit_test_rect_mouse->Inset(mouse_extend_);
diff --git a/ui/aura/window_tree_host.cc b/ui/aura/window_tree_host.cc
index a59d856aa5492b2e3c88125144e8df23f56fc002..c5f53964e4ce6b8c3d0b94be35cd65fdc598e36c 100644
--- a/ui/aura/window_tree_host.cc
+++ b/ui/aura/window_tree_host.cc
@@ -182,11 +182,109 @@ gfx::Transform WindowTreeHost::GetInverseRootTransform() const {
 }
 
 void WindowTreeHost::SetDisplayTransformHint(gfx::OverlayTransform transform) {
+  transform_hint_ = transform;
   if (compositor()->display_transform_hint() == transform)
     return;
 
   compositor()->SetDisplayTransformHint(transform);
-  UpdateCompositorScaleAndSize(GetBoundsInPixels().size());
+  OnHostResizedInPixels(GetBoundsInPixels().size());
+  CalculateInputTransform(transform);
+}
+
+gfx::Point WindowTreeHost::GetTransformedPoint(const gfx::Point& point) {
+  return input_transform_.MapPoint(point);
+}
+
+gfx::Rect WindowTreeHost::GetTransformedCompositorBounds(const gfx::Rect& new_bounds) {
+  if (IsPopup()) {
+    // Ensure that the compositor size is always the same as the original popup bounds.
+    return gfx::Rect(new_bounds.origin(), gfx::Size(original_popup_bounds_.width(), original_popup_bounds_.height()));
+  }
+
+  return GetTransformedWindowBounds(new_bounds);
+}
+
+gfx::Rect WindowTreeHost::GetTransformedWindowBounds(const gfx::Rect& new_bounds) {
+  gfx::Rect transformed_bounds(new_bounds);
+
+  if (IsPopup()) {
+    if (transform_hint_ == gfx::OVERLAY_TRANSFORM_ROTATE_CLOCKWISE_90 ||
+        transform_hint_ == gfx::OVERLAY_TRANSFORM_ROTATE_CLOCKWISE_270) {
+      // Make the window bounds equal the original popup bounds with the width and height swapped.
+      transformed_bounds.set_size(gfx::Size(original_popup_bounds_.height(), original_popup_bounds_.width()));
+    }
+  }
+  else {
+    gfx::Rect window_bounds = GetBoundsInPixels();
+    if ((transform_hint_ == gfx::OVERLAY_TRANSFORM_ROTATE_CLOCKWISE_90 ||
+         transform_hint_ == gfx::OVERLAY_TRANSFORM_ROTATE_CLOCKWISE_270) &&
+        window_bounds.height() && window_bounds.width()) {
+      float window_aspect_ratio = (float)window_bounds.width() / (float)window_bounds.height();
+      float new_aspect_ratio = (float)new_bounds.width() / (float)new_bounds.height();
+
+      // If the aspect ratio of the window and new bounds are the same it means that new_bounds
+      // has not already been transposed.
+      if ((window_aspect_ratio > 1.0 && new_aspect_ratio > 1.0) ||
+          (window_aspect_ratio < 1.0 && new_aspect_ratio < 1.0)) {
+        transformed_bounds.Transpose();
+      }
+    }
+  }
+  return transformed_bounds;
+}
+
+gfx::Rect WindowTreeHost::GetNonTransformedWindowBounds(const gfx::Rect& new_bounds) {
+  gfx::Rect non_transformed_bounds(new_bounds);
+  gfx::Rect window_bounds = GetBoundsInPixels();
+
+  if ((transform_hint_ == gfx::OVERLAY_TRANSFORM_ROTATE_CLOCKWISE_90 ||
+       transform_hint_ == gfx::OVERLAY_TRANSFORM_ROTATE_CLOCKWISE_270) &&
+       window_bounds.height() && window_bounds.width()) {
+    float window_aspect_ratio = (float)window_bounds.width() / (float)window_bounds.height();
+    float new_aspect_ratio = (float)new_bounds.width() / (float)new_bounds.height();
+
+    // If the aspect ratio of the window and new bounds are different it means that new_bounds
+    // has already been transposed, so we need to transpose it back.
+    if ((window_aspect_ratio > 1.0 && new_aspect_ratio < 1.0) ||
+        (window_aspect_ratio < 1.0 && new_aspect_ratio > 1.0)) {
+      non_transformed_bounds.Transpose();
+    }
+  }
+  return non_transformed_bounds;
+}
+
+void WindowTreeHost::CalculateInputTransform(gfx::OverlayTransform transform) {
+  input_transform_.MakeIdentity();
+  popup_transform_.MakeIdentity();
+
+  switch (transform) {
+    case gfx::OVERLAY_TRANSFORM_ROTATE_CLOCKWISE_90:
+      input_transform_.Translate(0, GetBoundsInPixels().width());
+      input_transform_.Rotate(-90);
+
+      popup_transform_.Translate(GetBoundsInPixels().width(), 0);
+      popup_transform_.Rotate(90);
+      break;
+
+    case gfx::OVERLAY_TRANSFORM_ROTATE_CLOCKWISE_180:
+      input_transform_.Translate(GetBoundsInPixels().width(), GetBoundsInPixels().height());
+      input_transform_.Rotate(180);
+
+      popup_transform_.Translate(GetBoundsInPixels().width(), GetBoundsInPixels().height());
+      popup_transform_.Rotate(180);
+      break;
+
+    case gfx::OVERLAY_TRANSFORM_ROTATE_CLOCKWISE_270:
+      input_transform_.Translate(GetBoundsInPixels().height(), 0);
+      input_transform_.Rotate(90);
+
+      popup_transform_.Translate(0, GetBoundsInPixels().height());
+      popup_transform_.Rotate(-90);
+      break;
+
+    default:
+      break;
+  }
 }
 
 gfx::Transform WindowTreeHost::GetRootTransformForLocalEventCoordinates()
@@ -205,19 +303,15 @@ gfx::Transform WindowTreeHost::GetInverseRootTransformForLocalEventCoordinates()
 
 void WindowTreeHost::UpdateCompositorScaleAndSize(
     const gfx::Size& new_size_in_pixels) {
-  gfx::Rect new_bounds(new_size_in_pixels);
-  if (compositor_->display_transform_hint() ==
-          gfx::OVERLAY_TRANSFORM_ROTATE_CLOCKWISE_90 ||
-      compositor_->display_transform_hint() ==
-          gfx::OVERLAY_TRANSFORM_ROTATE_CLOCKWISE_270) {
-    new_bounds.Transpose();
-  }
+  gfx::Rect orig_bounds(new_size_in_pixels);
+  gfx::Rect new_bounds = GetTransformedCompositorBounds(orig_bounds);
 
   // Allocate a new LocalSurfaceId for the new size or scale factor.
   window_->AllocateLocalSurfaceId();
   ScopedLocalSurfaceIdValidator lsi_validator(window());
   compositor_->SetScaleAndSize(device_scale_factor_, new_bounds.size(),
                                window_->GetLocalSurfaceId());
+  CalculateInputTransform(transform_hint_);
 }
 
 void WindowTreeHost::ConvertDIPToScreenInPixels(gfx::Point* point) const {
diff --git a/ui/aura/window_tree_host.h b/ui/aura/window_tree_host.h
index 6f4504050f5e2b6160bac9eb9c6029058c712fe6..c12544de5defed3ab28eef2461bc60623cdea90a 100644
--- a/ui/aura/window_tree_host.h
+++ b/ui/aura/window_tree_host.h
@@ -29,6 +29,7 @@
 #include "ui/events/platform_event.h"
 #include "ui/gfx/native_widget_types.h"
 #include "ui/gfx/overlay_transform.h"
+#include "ui/gfx/geometry/transform.h"
 
 namespace gfx {
 class Point;
@@ -126,6 +127,15 @@ class AURA_EXPORT WindowTreeHost : public ui::ImeKeyEventDispatcher,
 
   void SetDisplayTransformHint(gfx::OverlayTransform transform);
 
+  gfx::Point GetTransformedPoint(const gfx::Point& point);
+  gfx::Rect GetTransformedCompositorBounds(const gfx::Rect& new_bounds);
+  gfx::Rect GetTransformedWindowBounds(const gfx::Rect& new_bounds);
+  gfx::Rect GetNonTransformedWindowBounds(const gfx::Rect& new_bounds);
+  gfx::Transform GetPopupTransform() const { return popup_transform_; }
+  gfx::OverlayTransform GetDisplayTransformHint() const { return transform_hint_; }
+  bool IsPopup() const { return !original_popup_bounds_.IsEmpty(); }
+  void SetPopupBounds(const gfx::Rect& bounds) { original_popup_bounds_ = bounds; }
+
   // These functions are used in event translation for translating the local
   // coordinates of LocatedEvents. Default implementation calls to non-local
   // ones (e.g. GetRootTransform()).
@@ -436,6 +446,8 @@ class AURA_EXPORT WindowTreeHost : public ui::ImeKeyEventDispatcher,
   void OnFirstSurfaceActivation(ui::Compositor* compositor,
                                 const viz::SurfaceInfo& surface_info) override;
 
+  void CalculateInputTransform(gfx::OverlayTransform transform);
+
   // We don't use a std::unique_ptr for |window_| since we need this ptr to be
   // valid during its deletion. (Window's dtor notifies observers that may
   // attempt to reach back up to access this object which will be valid until
@@ -503,6 +515,12 @@ class AURA_EXPORT WindowTreeHost : public ui::ImeKeyEventDispatcher,
   // This is only used when occlusion tracking is always enabled.
   int video_capture_count_for_occlusion_tracking_ = 0;
 
+  gfx::OverlayTransform transform_hint_ = gfx::OVERLAY_TRANSFORM_NONE;
+
+  gfx::Transform input_transform_;
+  gfx::Transform popup_transform_;
+  gfx::Rect original_popup_bounds_;
+
   base::WeakPtrFactory<WindowTreeHost> weak_factory_{this};
 };
 
diff --git a/ui/views/widget/desktop_aura/desktop_native_widget_aura.cc b/ui/views/widget/desktop_aura/desktop_native_widget_aura.cc
index 006ed289a3fb9eefb8f697d353b4e85c5d7088be..d24ea2140e70ac0e03392e83e589a026ab2b3124 100644
--- a/ui/views/widget/desktop_aura/desktop_native_widget_aura.cc
+++ b/ui/views/widget/desktop_aura/desktop_native_widget_aura.cc
@@ -886,7 +886,11 @@ void DesktopNativeWidgetAura::SetBounds(const gfx::Rect& bounds) {
   if (!desktop_window_tree_host_) {
     return;
   }
-  desktop_window_tree_host_->SetBoundsInDIP(bounds);
+  gfx::Rect final_bounds(bounds);
+  if (host_ && host_->IsPopup()) {
+    final_bounds = host_->GetTransformedWindowBounds(bounds);
+  }
+  desktop_window_tree_host_->SetBoundsInDIP(final_bounds);
 }
 
 void DesktopNativeWidgetAura::SetBoundsConstrained(const gfx::Rect& bounds) {
